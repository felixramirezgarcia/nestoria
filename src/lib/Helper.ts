// Native.
import * as fs from 'fs';
import * as path from 'path';

// Package.
import * as _ from 'lodash';

// Code.
export class Helper {
  public static readFileLines(filePath: string): string[] {
    return fs
      .readFileSync(path.resolve(filePath))
      .toString()
      .split('\n');
  }

  public static deepOmit(object: any, keysToOmit: string[]) {
    const keysToOmitIndex = _.keyBy(Array.isArray(keysToOmit) ? keysToOmit : [keysToOmit]);

    function omitFromObject(obj: any) {
      return _.transform(obj, (result: any, value: any, key: string) => {
        if (key in keysToOmitIndex) {
          return;
        }
        result[key] = _.isObject(value) ? omitFromObject(value) : value;
      });
    }

    return omitFromObject(object);
  }
}
