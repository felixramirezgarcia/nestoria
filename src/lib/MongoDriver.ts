// Package.
import * as mongoose from 'mongoose';
import * as debug from 'debug';

// Internals.
import { DBSetup } from './Config';

// Code.
const debugVerbose = debug('driver:verbose:mongo');
const debugError = debug('driver:error:mongo');

export type MongoDriverParams = DBSetup;

export interface MongoModelParams {
  name: string;
  schema?: mongoose.Schema;
  collection?: string;
  skipInit?: boolean;
}

export abstract class MongoModel<T extends mongoose.Document> {
  protected modelParams: MongoModelParams;
  protected model: mongoose.Model<T>;

  constructor() {
    this.modelParams = {
      name: 'Collection',
      skipInit: true,
    };
  }

  get params(): MongoModelParams {
    return this.modelParams;
  }

  protected abstract attachToDriver(driver: MongoDriver): void;
}

export default class MongoDriver {
  private static instance: MongoDriver;
  private params: MongoDriverParams;
  private connection: mongoose.Mongoose;

  private constructor(config: MongoDriverParams) {
    this.params = config;
    this.params.dbHost = config.dbHost ? config.dbHost : 'localhost';
    this.params.dbPort = config.dbPort ? config.dbPort : 27017;
    this.params.dbUser = config.dbUser ? config.dbUser : '';
    this.params.dbPassword = config.dbPassword ? config.dbPassword : '';
    this.registerEvents();
  }

  public static getInstance(config: MongoDriverParams): MongoDriver {
    if (!MongoDriver.instance) {
      MongoDriver.instance = new MongoDriver(config);
    }

    return MongoDriver.instance;
  }

  public async connect(): Promise<mongoose.Mongoose> {
    if (this.connection == null) {
      const options: mongoose.ConnectionOptions = {
        keepAlive: true,
        connectTimeoutMS: 30000,
        // If not connected, return errors immediately rather than waiting for reconnect
        bufferMaxEntries: 0,
        // Never stop trying to reconnect
        reconnectTries: Number.MAX_VALUE,
        useNewUrlParser: true,
      };

      try {
        const userPassBind =
          this.params.dbUser && this.params.dbPassword ? `${this.params.dbUser}:${this.params.dbPassword}@` : '';

        if (process.env.MONGO_DEBUG) {
          mongoose.set('debug', true);
        }

        this.connection = await mongoose.connect(
          `mongodb://${userPassBind}${this.params.dbHost}:${this.params.dbPort}/${this.params.dbName}`,
          options
        );
      } catch (err) {
        debugError(err);
      }
    }
    return this.connection;
  }

  public attachModel<T extends mongoose.Document>(model: MongoModel<T>): mongoose.Model<T> {
    const m: mongoose.Model<T> = this.connection.connection.model<T>(
      model.params.name,
      model.params.schema,
      model.params.collection
    );

    return m;
  }

  public async disconnect(cb?: () => void): Promise<void> {
    if (this.connection) {
      if (cb) {
        await this.connection.disconnect(cb);
      } else {
        await this.connection.disconnect();
      }
      this.connection = null;
    }
  }

  public getConnector(): mongoose.Mongoose {
    return this.connection;
  }

  private registerEvents(): void {
    mongoose.connection.on('connected', () => {
      debugVerbose(`Mongoose connected to ${this.params.dbName}`);
    });
    mongoose.connection.on('disconnected', () => {
      debugVerbose(`Mongoose disconnected from ${this.params.dbName}`);
    });
    mongoose.connection.on('error', (err: Error) => {
      debugVerbose(`Mongoose has errors ${err.stack}`);
    });
  }
}
