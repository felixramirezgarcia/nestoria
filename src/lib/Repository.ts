// Package.
import { Document, Model, Types } from 'mongoose';
import { InsertWriteOpResult } from 'mongodb';

// Code.
export interface Read<T> {
  retrieve: () => Promise<T[]>;
  findById: (id: string) => Promise<T>;
  findOne(cond?: any): Promise<T>;
  find(cond: any, fields: any, options: any): Promise<T[]>;
}

export interface Write<T> {
  create: (item: T) => Promise<T>;
  bulkCreate: (items: T[]) => Promise<InsertWriteOpResult>;
  update: (id: string, item: T) => Promise<T>;
  delete: (id: string) => Promise<T>;
}

export interface BulkDelete {
  n: number;
  ok: number;
}

export class RepositoryBase<T extends Document> implements Read<T>, Write<T> {
  private _model: Model<Document>;

  constructor(schemaModel: Model<Document>) {
    this._model = schemaModel;
  }

  public async create(item: T): Promise<T> {
    let output = null;

    try {
      output = (await this._model.create(item)) as T;
    } catch (err) {
      throw err;
    }

    return output;
  }

  public async bulkCreate(items: T[]): Promise<InsertWriteOpResult> {
    let output = null;

    try {
      output = await this._model.collection.insertMany(items);
    } catch (err) {
      throw err;
    }

    return output;
  }

  public async update(id: string, item: T): Promise<T> {
    let output = null;

    try {
      const _id: Types.ObjectId = this.toObjectId(id);
      await this._model.update({ _id }, item);
      output = (await this._model.findById({ _id })) as T;
    } catch (err) {
      throw err;
    }

    return output;
  }

  public async delete(id: string): Promise<T> {
    let output = null;

    try {
      output = (await this._model.findById(id)) as T;
      await this._model.remove({ _id: id });
    } catch (err) {
      throw err;
    }

    return output;
  }

  public async retrieve(): Promise<T[]> {
    let output: T[] = null;

    try {
      const conditions = {};
      output = (await this._model.find(conditions)) as T[];
    } catch (err) {
      throw err;
    }

    return output;
  }

  public async findById(_id: string): Promise<T> {
    let output = null;

    try {
      output = (await this._model.findById(_id)) as T;
    } catch (err) {
      throw err;
    }

    return output;
  }

  public async findOne(cond?: any): Promise<T> {
    let output = null;

    try {
      output = (await this._model.findOne(cond)) as T;
    } catch (err) {
      throw err;
    }

    return output;
  }

  public async find(cond?: any, options?: any): Promise<T[]> {
    let output = null;

    try {
      output = (await this._model.find(cond, options)) as T[];
    } catch (err) {
      throw err;
    }

    return output;
  }

  private toObjectId(_id: string): Types.ObjectId {
    return Types.ObjectId.createFromHexString(_id);
  }
}
