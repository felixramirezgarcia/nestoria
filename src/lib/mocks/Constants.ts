// Code.
export const keysToOmit = ['createdAt', 'updatedAt', '__v', '_id'];
