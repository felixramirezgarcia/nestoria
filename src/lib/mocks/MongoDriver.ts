// Internal.
import { MongoDriverParams } from '../MongoDriver';

// Code.
export const mockConnection: MongoDriverParams = {
  dbName: 'test',
  dbHost: 'localhost',
  dbPort: 27017,
};
