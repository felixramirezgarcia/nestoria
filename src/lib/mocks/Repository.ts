// Package.
import { Schema, Document, SchemaDefinition, SchemaOptions, Model } from 'mongoose';

// Internal.
import MongoDriver, { MongoModel } from '../MongoDriver';
import { RepositoryBase } from '../Repository';

// Code.
export interface TestDocument extends Document {
  test: number;
  createdAt?: Date;
  updatedAt?: Date;
}

export const mockedTestItem = {
  test: 1,
} as TestDocument;

const definitions: SchemaDefinition = {
  test: { type: Number, required: true },
};
const options: SchemaOptions = {
  timestamps: true,
};

const schema: Schema = new Schema(definitions, options);

export class TestModel extends MongoModel<TestDocument> {
  public repo: TestRepository;

  constructor() {
    super();
    this.modelParams.name = 'Test';
    this.modelParams.schema = schema;
    this.modelParams.collection = 'test';
  }

  public attachToDriver(driver: MongoDriver) {
    this.model = driver.attachModel<TestDocument>(this);
    this.repo = TestRepository.getInstance(this.model);
  }
}

export class TestRepository extends RepositoryBase<TestDocument> {
  private static instance: TestRepository;

  private constructor(model: Model<TestDocument>) {
    super(model);
  }

  public static getInstance(model: Model<TestDocument>): TestRepository {
    if (!TestRepository.instance) {
      TestRepository.instance = new TestRepository(model);
    }

    return TestRepository.instance;
  }
}
