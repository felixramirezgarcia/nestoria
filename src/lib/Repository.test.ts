// Internal.
import MongoDriver from './MongoDriver';
import { Helper } from './Helper';
import { keysToOmit } from './mocks/Constants';

// Mocks.
import { mockConnection } from './mocks/MongoDriver';
import { TestModel, TestRepository, TestDocument, mockedTestItem } from './mocks/Repository';

// Code.
describe('Repository', () => {
  let driver: MongoDriver;
  let repository: TestRepository;
  let lastEntry: TestDocument;

  beforeAll(async () => {
    driver = MongoDriver.getInstance(mockConnection);
    await driver.connect();
    const model = new TestModel();
    model.attachToDriver(driver);
    repository = model.repo;
  });

  afterAll(async () => {
    await driver.disconnect();
  });

  describe('deleteAll', () => {
    it('should delete all the entries in the collection', async () => {
      const result = await repository.deleteAll();
      expect(result.ok).toBeGreaterThan(0);
    });
  });

  describe('create', () => {
    it('should create a new entry', async () => {
      const result = await repository.create(mockedTestItem);
      lastEntry = result;
      expect(Helper.deepOmit(result, keysToOmit)).toMatchSnapshot();
    });

    it('should create 2 new entries at once', async () => {
      const entries = [];
      for (let i = 2; i <= 3; i++) {
        entries.push({
          test: i,
        } as TestDocument);
      }

      const result = await repository.bulkCreate(entries);
      expect(result.insertedCount).toBe(2);
    });
  });

  describe('find', () => {
    it('should retrieve all the existent entries', async () => {
      const result = await repository.retrieve();
      expect(Helper.deepOmit(result, keysToOmit)).toMatchSnapshot();
    });

    it('should find whatever according the query conditions', async () => {
      const result = await repository.find({ test: { $gt: 0 } });
      expect(Helper.deepOmit(result, keysToOmit)).toMatchSnapshot();
    });

    it('should find an entry using its attributes', async () => {
      const result = await repository.findOne({ test: mockedTestItem.test });
      expect(Helper.deepOmit(result, keysToOmit)).toMatchSnapshot();
    });

    it('should find an entry by its id', async () => {
      const result = await repository.findById(lastEntry._id);
      expect(Helper.deepOmit(result, keysToOmit)).toMatchSnapshot();
    });
  });

  describe('update', () => {
    it('should update an entry by its id', async () => {
      const doc = {
        test: 43,
      } as TestDocument;
      const result = await repository.update(lastEntry.id, doc);
      expect(Helper.deepOmit(result, keysToOmit)).toMatchSnapshot();
    });
  });

  describe('delete', () => {
    it('should delete the updated entry', async () => {
      const result = await repository.delete(lastEntry._id);
      expect(Helper.deepOmit(result, keysToOmit)).toMatchSnapshot();
    });

    it('should flush all the collection', async () => {
      const result = await repository.deleteAll();
      expect(result.n).toBe(2);
    });
  });
});
