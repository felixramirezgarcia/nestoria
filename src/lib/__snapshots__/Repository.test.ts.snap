// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repository create should create a new entry 1`] = `
Object {
  "test": 1,
}
`;

exports[`Repository delete should delete the updated entry 1`] = `
Object {
  "test": 43,
}
`;

exports[`Repository find should find an entry by its id 1`] = `
Object {
  "test": 1,
}
`;

exports[`Repository find should find an entry using its attributes 1`] = `
Object {
  "test": 1,
}
`;

exports[`Repository find should find whatever according the query conditions 1`] = `
Array [
  Object {
    "test": 1,
  },
  Object {
    "test": 2,
  },
  Object {
    "test": 3,
  },
]
`;

exports[`Repository find should retrieve all the existent entries 1`] = `
Array [
  Object {
    "test": 1,
  },
  Object {
    "test": 2,
  },
  Object {
    "test": 3,
  },
]
`;

exports[`Repository update should update an entry by its id 1`] = `
Object {
  "test": 43,
}
`;
